### DISCUSSION TITLE
blog/constrained-optimization-lagrangian-projected-gradient-and-penalty-function-method/

### DISCUSSION THREAD

**Discussion by giscus** [jeremy-feng/giscus#13] [Announcements]
# blog/constrained-optimization-lagrangian-projected-gradient-and-penalty-function-method/

使用 Lagrangian 乘子法、投影梯度算法、罚函数法求解有约束的优化问题。

https://fengchao.pro/blog/constrained-optimization-lagrangian-projected-gradient-and-penalty-function-method/

**Comment by ZhangQuan-hub** [comment]
大家不用试了，根本跑不通

**Reply by jeremy-feng** [reply]
请问哪一部分跑不通呀

**Comment by wuzhe521** [comment]
罚函数的梯度好像整错了，应该是对 h^2(X)求偏导，程序里是对h(X)求导

**Reply by wuzhe521** [reply]
return np.array([2 * h(X) * X[0], 2  * h(X) * X[1]])

**Reply by jeremy-feng** [reply]
感谢指正🙏

代码已修改。

**Reply by jeremy-feng** [reply]
另外，结果中的每一项应该有两个 $2 *$，分别是 $h^2(x)$ 和 $x^2$ 的导数带来的。
